File: String/541.反转字符串-ii.cpp
#include <iostream>

using namespace std;
/*
 * @lc app=leetcode.cn id=541 lang=cpp
 *
 * [541] 反转字符串 II
 */

// @lc code=start
class Solution
{
public:
    // 反转字符串的函数
    void reverse(string &str, int start, int end)
    {

        for (int i = start, j = end; i < j; i++, j--)
        {
            int temp = str[i];
            str[i] = str[j];
            str[j] = temp;
        }
    }

    string reverseStr(string s, int k)
    {
        int size = s.size();

        for (int i = 0; i < size; i += 2 * k)
        {
            int endSize = size - i;

            if (endSize >= k)
            {
                // 只反转前k个字符
                reverse(s, i, i + k - 1);
            }
            else
            {
                // 反转剩下的全部
                reverse(s, i, size - 1);
            }
        }

        return s;
    }
};
// @lc code=end

File: String/541.反转字符串-ii.ts
/*
 * @lc app=leetcode.cn id=541 lang=typescript
 *
 * [541] 反转字符串 II
 */

// @lc code=start
function reverseString(s: string[], left, right): void {
  while (left < right) {
    let temp = s[left]
    s[left] = s[right]
    s[right] = temp
    left++
    right--
  }
};

function reverseStr(s: string, k: number): string {
  let stringArr = s.split("")
  // 每次走2k
  for (let i = 0; i < s.length; i += 2 * k) {
    let left = i
    let right = 0

    // 剩余个数超过k个，依然翻转前k个
    if (s.length - i > k) {
      right = i + k - 1
    } else {
      // 不足k个，翻转全部
      right = s.length - 1
    }
    // 反转数组
    reverseString(stringArr, left, right)
  }

  return stringArr.join("")
};
// @lc code=end

